src/plots/cartesian/axes.js

```diff
@@ -777,20 +781,31 @@ axes.calcTicks = function calcTicks(ax, opts) {
             var A = tickVals[a].value;
             var B = tickVals[b].value;
 
-            var delta = definedDelta || Math.abs(B - A);
-            if(delta >= ONEDAY * 365) { // Years could have days less than ONEAVGYEAR period
-                v += ONEAVGYEAR / 2;
-            } else if(delta >= ONEAVGQUARTER) {
-                v += ONEAVGQUARTER / 2;
-            } else if(delta >= ONEMINMONTH) { // Months could have days less than ONEAVGMONTH period
-                var actualDelta = Math.abs(B - A);
+            var actualDelta = Math.abs(B - A);
+            var delta = definedDelta || actualDelta;
+            if(delta >= ONEMINYEAR) {
+                if(actualDelta >= ONEMINYEAR && actualDelta <= ONEMAXYEAR) {
+                    v += actualDelta / 2;
+                } else {
+                    v += ONEAVGYEAR / 2;
+                }
+            } else if(delta >= ONEMINQUARTER) {
```

Comment:
```suggestion
            } else if(definedDelta && (delta >= ONEMINQUARTER)) {
```
These only apply if we got the delta from `tickformat`; otherwise we can still get 3-month `dtick` but we're using default formatting so the shift should still be half a month.

Same for weeks, default formatting has days so the shift should be half a day.

Reply:
Not necessarily. Better to improve positioning for the cases even if the `delta` was not coming from `tickformat`.

Reply:
But without `tickformat`, in the case of quarter or week `delta` the shift should NOT be half a quarter/week because that's not what we're labeling: we're labeling months/days

Reply:
Also another case to consider: we have `tickformat` defined, but the user zoomed in so now `dtick` is smaller than the increment described by `tickformat`. Then there will be multiple identical ticks, but we should offset by half the actual spacing rather than half the `definedDelta`.

And yet another - not new here, but what if the user uses `tickformatstops`? Seems like we should use `axes.getTickFormat(ax)` rather than `ax.tickformat` to determine `definedDelta`

Reply:
🤯

Reply:
Revised in https://github.com/plotly/plotly.js/pull/5089/commits/ea6eca1998614364a1554b53a7e6403621b435ec and https://github.com/plotly/plotly.js/pull/5089/commits/37a08402b71de6d8a80ada549f8d606eb34cdea6.